src/engine/database/mod.rs

```diff
@@ -134,6 +135,86 @@ pub trait DatabaseEndpoint {
     async fn pool(&self) -> Result<DatabasePool, Error>;
 }
 
+/// Represents the different type of comparison match operations
+#[derive(Clone, Debug)]
+pub enum Operation {
+    EQ,
+    CONTAINS,
+    IN,
+    GT,
+    GTE,
+    LT,
+    LTE
+}
+
+/// Struct representing a value comparison. In query operations, visitors take provided 
+/// operation/value nested map and converted them into a `Comparison` struct and pass
+/// it on the database-specific transaction for use in creating match portion of queries. 
+#[derive(Clone, Debug)]
+pub struct Comparison {
+    operation: Operation,
+    operand: Value,
+    negated: bool
+}
+
+impl Comparison {
+
+    pub fn new(operation: Operation, negated: bool, operand: Value) -> Self {
+        Comparison {
+            operation,
+            operand,
+            negated
+        }
+    }
+
+    pub fn default(v: Value) -> Self {
+        Self::new(Operation::EQ, false, v)
+    }
+
+}
+
+impl TryFrom<Value> for Comparison {
+
+    type Error = Error;
+
+    fn try_from(v: Value) -> Result<Comparison, Error> {
+        Ok(match v {
+            Value::String(_) => Comparison::default(v),
+            Value::Int64(_) => Comparison::default(v),
+            Value::Float64(_) => Comparison::default(v),
+            Value::Bool(_) => Comparison::default(v),
+            Value::Map(mut m) => {
```

Comment:
So, correct me if I'm wrong, but basically this lets you use the visitors to follow the input structure down to the point that you want to handle query values that may have comparisons, and then you just throw that input `Value` map at this `try_from` block and it converts everything for you?  If I understand this correctly, wow... that's more ergonomic than I expected would be possible.

Reply:
Yup! I was pleasantly surprised that it turned out to be this clean.